File: baekjoon/1213.cpp
#include <iostream>
#include <algorithm>
using namespace std;

// 팰린드롬 만들기

string name;
int alphabet[26];
char mid = '0';
bool isPossible() {
    
    for(int i = 0; i<name.size(); i++)
        alphabet[name[i] - 'A']++;
    
    int flag = 0;
    for(int i = 0; i<26; i++) {
        if(alphabet[i] % 2 == 1) { // 홀수개의 문자열이 있다면
            flag++; // 개수 카운트
            mid = i + 'A'; // 해당 문자 저장
        }
        if(flag > 1) return false;
    }
    
    return true;
}

int main() {
    
    cin >> name;
    
    sort(name.begin(), name.end());
    
    if(isPossible()) {
        string answer = "";
        for(int i = 0; i<26; i++) {
            for(int j = 0; j<alphabet[i]/2; j++)
                answer += i + 'A';
        }
        
        cout << answer;
        if(mid != '0')
            cout << mid;
        reverse(answer.begin(), answer.end());
        cout << answer;
    }
    else cout << "I'm Sorry Hansoo";
    
    return 0;
}

File: baekjoon/1213.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*
import kotlin.collections.ArrayList

// 팰린드롬 만들기

fun main() {
    val bufferedReader = BufferedReader(InputStreamReader(System.`in`))
    val bufferedWriter = BufferedWriter(OutputStreamWriter(System.out))
    val token = StringTokenizer(bufferedReader.readLine())
    val name: String = token.nextToken().toString()
    val alphabet: IntArray = IntArray(26) { 0 }
    var middleIndex: Int = -1

    fun isPossible(name: CharArray): Boolean {
        var oddCount = 0

        for (i in name.indices)
            alphabet[name[i].toInt().minus('A'.toInt())]++

        for (i in 0 until 26)
            if (alphabet[i] % 2 == 1) {
                oddCount++
                middleIndex = i
            }

        if (name.size % 2 == 1 && oddCount == 1)
            return true

        if (name.size % 2 == 0 && oddCount == 0)
            return true

        return false
    }


    if (isPossible(name.toCharArray())) {
        var half: String = ""

        for (i in 0 until 26) {
            for (j in 0 until alphabet[i] / 2)
                half += "${i.plus('A'.toInt()).toChar()}"
        }
        bufferedWriter.write(half)

        if (middleIndex != -1)
            bufferedWriter.write("${middleIndex.plus('A'.toInt()).toChar()}")

        bufferedWriter.write(half.reversed())
    } else {
        bufferedWriter.write("I'm Sorry Hansoo")
    }

    bufferedWriter.flush()
    bufferedWriter.close()


}